Move orphan processing to ActivateBestChain

TheBlueMatt opened the pull request:
Based on #8928 because its the only way I can run tests.
This will conflict a bit with #8865, but that's easy to resolve, and I'd rather get some eyeballs on this sooner rather than later, so am leaving it freestanding. Once #8865 goes in the orphan processing will go into the CValidationInterface which manges "net processing" stuff.

This further decouples "main" and "net" processing logic by moving
orphan processing out of the chain-connecting cs_main lock and
into its own cs_main lock, beside all of the other chain callbacks.

Once further decoupling of net and main processing logic occurs,
orphan handing should move to its own lock, out of cs_main.

Note that this will introduce a race if there are any cases where
we assume the orphan map to be consistent with the current chain
tip, however I am confident there is no such case (ATMP will fail
without DoS score in all such cases).

TheBlueMatt committed:
Move orphan processing to ActivateBestChain

This further decouples "main" and "net" processing logic by moving
orphan processing out of the chain-connecting cs_main lock and
into its own cs_main lock, beside all of the other chain callbacks.

Once further decoupling of net and main processing logic occurs,
orphan handing should move to its own lock, out of cs_main.

Note that this will introduce a race if there are any cases where
we assume the orphan map to be consistent with the current chain
tip, however I am confident there is no such case (ATMP will fail
without DoS score in all such cases).

sipa commented on src/main.cpp:
```diff
@@ -3068,10 +3048,38 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
         if(connman)
             connman->SetBestHeight(nNewHeight);
 
+
```
Useful.

sipa commented on src/main.cpp:
```diff
@@ -3068,10 +3048,38 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
 
+        // Remove orphan transactions with cs_main
+        {
+            LOCK(cs_main);
```
I hope we can separate nodestate and other message-processing related variables from cs_main...

theuni replied:
@sipa Absolutely, I'm planning on doing that after the dust settles here, unless @TheBlueMatt already has it queued up. I'm planning on moving lots of what's currently in CNode to CNodeState, so breaking out the locks is a prerequisite for that.

TheBlueMatt replied:
Dont have it queued up, but, indeed, thats an obvious change after we split main.cpp.

TheBlueMatt committed:
Erase orphans per-transaction instead of per-block

TheBlueMatt committed:
Move orphan-conflict removal from main logic into a callback

This makes the orphan map a part of net-processing logic instead
of main logic.

TheBlueMatt commented:
Closing for now. Will rebase on #9075 when its merged.

TheBlueMatt commented:
Rebased on #9075...this now forms the last pull before main.cpp is split clean in two :).

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/8930/commits/2dd652afe577a09d6f70db0e79b8f6c0706dd1b4

rebroad commented:
What is the rationale behind this refactoring please? What future changes is it enabling? Or what is it fixing?

laanwj commented:
@rebroad 

> This further decouples "main" and "net" processing logic by moving
> orphan processing out of the chain-connecting cs_main lock and
> into its own cs_main lock, beside all of the other chain callbacks.

Decoupling unrelated logic to be able to split up main into validation and network handling parts.

sdaftuar commented:
@TheBlueMatt  One thing we don't currently do, but probably should, is try to reaccept orphan transactions to the mempool if their missing inputs appear in a block.  After this PR, would you suggest we just do that inside `PeerLogicValidation::SyncTransaction()`, or would that be problematic?

sipa commented:
utACK. You could get rid of the BOOST_FOREACH along the way, if you wish.

TheBlueMatt commented:
@sdaftuar that seems reasonable, yea. Not for this PR, though, indeed.

sdaftuar commented:
utACK

TheBlueMatt commented on src/main.cpp:
```diff
@@ -4744,6 +4724,30 @@ PeerLogicValidation::PeerLogicValidation(CConnman* connmanIn) : connman(connmanI
 }
 
+void PeerLogicValidation::SyncTransaction(const CTransaction& tx, const CBlockIndex* pindex, int nPosInBlock) {
+    LOCK(cs_main);
```
Hmmmmm...good point...for corectness I'll add that check to this PR, but I think we probably actually want to remove it in a future one since I believe it would allow us to clean up orphan cleanup logic in the ::TX message handling logic.

theuni replied:
Yep, makes sense. But for now, it can at least be a short-circuit for returning without locking cs_main (not sure if there's any unlocked caller who would benefit, though).

TheBlueMatt replied:
Done

morcos replied:
@TheBlueMatt Your comment confuses me a bit.  Wouldn't this be broken without that check?

morcos replied:
I agree, nice to move all logic to this function. 
But for now, we don't actually remove orphan txs which conflict with things accepted to mempool.  We could, however, move recursive processing of orphan txs that depended on mempool accepted txs here, I agree.
Also there are other places SyncTransaction is called, such as for disconnected block txs, where it would be incorrect to remove orphan txs which conflict with txs from disconnected blocks.
In any case, the code is correct now, I just want to be sure we think about all the ways SyncTransaction is called when we change it in the future.

theuni commented:
utACK d2b88f97a1235057290f9e8ca0bf11437ba919f8

rebroad commented on src/main.cpp:
```diff
@@ -4744,6 +4724,34 @@ PeerLogicValidation::PeerLogicValidation(CConnman* connmanIn) : connman(connmanI
+        for (auto mi = itByPrev->second.begin(); mi != itByPrev->second.end(); ++mi) {
+            const CTransaction& orphanTx = (*mi)->second.tx;
+            const uint256& orphanHash = orphanTx.GetHash();
+            vOrphanErase.push_back(orphanHash);
```
Why create vOrphanErase at all? Why not simply run the EraseOrphanTx() here instead?
